### PR TITLE
cpu/samd5x: CPU init fixes

### PR DESCRIPTION
**Pull request by benpicco**
There is still a bug with the samd5x code:
I first sourced the DPLL0 from XOSC32K where it doesn't need a reference clock, so I deleted the reference clock configuration.

Then I wanted to have more flexibility, e.g. running the system entirely from the ULPOSC32K without the need for an external oscillator, so I sourced the DPLL0 from the 32kHz GCLK1 but forgot to add back the reference clock setup (this can be GCLK1 too).

Now for some reason this worked fine until I flashed an image with riotboot. A colleague noted that this used to work before, so I began searching as to what changed in the meantime. I could narrow it down to the dpll setup and indeed with this patch applied, riotboot can load an image again.

### CODE REVIEW

File: cpu/samd5x/cpu.c [hunk 1]
```diff
@@ -57,6 +57,12 @@ static void fdpll0_init(uint32_t f_cpu)
                        | OSCCTRL_DPLLCTRLB_WUF
                        | OSCCTRL_DPLLCTRLB_LBYPASS;
 
+    GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg = GCLK_PCHCTRL_GEN(1) | GCLK_PCHCTRL_CHEN;
```
Comment (benpicco) [inline cpu/samd5x/cpu.c]: I already did this in line 51 in the same function oO

File: cpu/samd5x/cpu.c [hunk 2]
```diff
@@ -49,6 +49,10 @@ static void fdpll0_init(uint32_t f_cpu)
     const uint32_t LDR = ((f_cpu << 5) / 32768);
 
     GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg = GCLK_PCHCTRL_GEN(1) | GCLK_PCHCTRL_CHEN;
+    GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL032K].reg = GCLK_PCHCTRL_GEN(1) | GCLK_PCHCTRL_CHEN;
```
Comment (dylad) [inline cpu/samd5x/cpu.c]: From what I understand, this clock is for the lock timer timeout and is optional. Do we really need this ??

File: cpu/samd5x/cpu.c [hunk 3]
```diff
@@ -49,6 +49,10 @@ static void fdpll0_init(uint32_t f_cpu)
...
+
+    while (!(GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg & GCLK_PCHCTRL_CHEN)) {}
```
Comment (dylad) [inline cpu/samd5x/cpu.c]: Could you please move this line just after `GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg = GCLK_PCHCTRL_GEN(1) | GCLK_PCHCTRL_CHEN;` ?

**Comment by benpicco** [comment]
Ok I think now it should be right.

That is, `make -C tests/riotboot riotboot/flash-combined-slot0` now works on both same54-xpro and our custom samd51 board, or `cpu_init()` survives being called twice.

Another gotcha is that DPLL must run with at least 96 MHz, so setting a lower CPU frequency invokes undefined behavior unless a divider is applied.
But this is for another PR.

File: cpu/samd5x/cpu.c [hunk 4]
```diff
@@ -41,21 +40,29 @@ static void dfll_init(void)
...
+
+    /* set DPLL clock source */
     GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg = GCLK_PCHCTRL_GEN(1) | GCLK_PCHCTRL_CHEN;
+    while (!(GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg & GCLK_PCHCTRL_CHEN)) {}
 
     OSCCTRL->Dpll[0].DPLLRATIO.reg = OSCCTRL_DPLLRATIO_LDRFRAC(LDR & 0x1F)
-                       | OSCCTRL_DPLLRATIO_LDR((LDR >> 5) - 1);
+                                   | OSCCTRL_DPLLRATIO_LDR((LDR >> 5) - 1);
 
+    /* Without LBYPASS, startup takes very long, reson unknown. */
```
Comment (dylad) [inline cpu/samd5x/cpu.c]: Reason is sillicon bug,
check errata [there](http://ww1.microchip.com/downloads/en/DeviceDoc/SAM-D5X-E5X-Family-Silicon-Errata-and-Data-Sheet-Clarification-80000748F.pdf). section 2.13

File: cpu/samd5x/cpu.c [hunk 5]
```diff
@@ -41,21 +40,29 @@ static void dfll_init(void)
...
     OSCCTRL->Dpll[0].DPLLCTRLB.reg = OSCCTRL_DPLLCTRLB_REFCLK_GCLK
-                       | OSCCTRL_DPLLCTRLB_WUF
```
Comment (dylad) [inline cpu/samd5x/cpu.c]: I would be better to not remove this regarding errata stuff.

File: cpu/samd5x/cpu.c (hunk 4 above)
Comment (benpicco) [reply cpu/samd5x/cpu.c]: Ah thank you! I should read the errata sheet more often

File: cpu/samd5x/cpu.c (hunk 5 above)
Comment (benpicco) [reply cpu/samd5x/cpu.c]: It's back - with the errata now it has a proper justification to be there.

**Comment by dylad** [comment]
Thanks for adding reference to the errata. I'll test it asap.

**Comment by dylad** [comment]
@benpicco could you rebase please ?

**Comment by benpicco** [comment]
Ok.

File: cpu/samd5x/cpu.c [hunk 6]
```diff
@@ -41,21 +40,30 @@ static void dfll_init(void)
...
+    OSCCTRL->Dpll[0].DPLLCTRLA.bit.ENABLE = 0;
+    while (OSCCTRL->Dpll[0].DPLLSYNCBUSY.reg) {}
+
+    /* set DPLL clock source */
     GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg = GCLK_PCHCTRL_GEN(1) | GCLK_PCHCTRL_CHEN;
+    while (!(GCLK->PCHCTRL[OSCCTRL_GCLK_ID_FDPLL0].reg & GCLK_PCHCTRL_CHEN)) {}
 
     OSCCTRL->Dpll[0].DPLLRATIO.reg = OSCCTRL_DPLLRATIO_LDRFRAC(LDR & 0x1F)
-                       | OSCCTRL_DPLLRATIO_LDR((LDR >> 5) - 1);
+                                   | OSCCTRL_DPLLRATIO_LDR((LDR >> 5) - 1);
```
[the remark this answers is not in the document]
Comment (benpicco) [reply cpu/samd5x/cpu.c]: Shouldn't one such loop after the enable be enough?

**Pull request merged**